src/sentry/models/projectownership.py

```diff
@@ -48,54 +51,65 @@ def get_owners(cls, project_id, data):
                 project_id=project_id,
             )
 
+        rules = []
         if ownership.schema is not None:
             for rule in load_schema(ownership.schema):
                 if rule.test(data):
-                    # This is O(n) to resolve, but should be fine for now
-                    # since we don't even explain that you can use multiple
-                    # let alone a number that would be potentially abusive.
-                    owners = []
-                    for o in rule.owners:
-                        try:
-                            owners.append(resolve_actor(o, project_id))
-                        except UnknownActor:
-                            continue
-                    return owners, rule.matcher
+                    rules.append(rule)
 
-        owners = cls.Everyone if ownership.fallthrough else []
-        return owners, None
+        if not rules:
+            return cls.Everyone if ownership.fallthrough else [], None
 
+        owners = {o for rule in rules for o in rule.owners}
 
-class UnknownActor(Exception):
-    pass
+        return filter(None, resolve_actors(owners, project_id).values()), rules
 
 
-def resolve_actor(owner, project_id):
-    """ Convert an Owner object into an Actor """
+def resolve_actors(owners, project_id):
+    """ Convert a list of Owner objects into a dictionary
+    of {Owner: Actor} pairs. Actors not identified are returned
+    as None. """
     from sentry.api.fields.actor import Actor
+    from sentry.ownership.grammar import Owner
     from sentry.models import User, Team
 
-    if owner.type == 'user':
-        try:
-            user_id = User.objects.filter(
-                email__iexact=owner.identifier,
-                is_active=True,
-                sentry_orgmember_set__organizationmemberteam__team__projectteam__project_id=project_id,
-            ).values_list('id', flat=True)[0]
-        except IndexError:
-            raise UnknownActor
+    if not owners:
+        return []
 
-        return Actor(user_id, User)
+    users, teams = [], []
 
-    if owner.type == 'team':
-        try:
-            team_id = Team.objects.filter(
+    owners_to_actors = {o: None for o in owners}
+
+    for owner in owners:
+        if owner.type == 'user':
+            users.append(owner)
+        elif owner.type == 'team':
+            teams.append(owner)
```

Comment:
I think it'd be good to have an else block here that raises or warns in case you get anything you weren't expecting.

Reply:
💯